core/src/main/java/jenkins/telemetry/Telemetry.java

```diff
@@ -0,0 +1,196 @@
+     * Will be checked in Jenkins to not collect outside the defined time span.
+     * This does not have to be precise enough for time zones to be a consideration.
+     *
+     * @return collection start date
+     */
+    @Nonnull
+    public abstract LocalDate getStart();
+
+    /**
+     * End date for the collection.
+     * Will be checked in Jenkins to not collect outside the defined time span.
+     * This does not have to be precise enough for time zones to be a consideration.
+     *
+     * @return collection end date
+     */
+    @Nonnull
+    public abstract LocalDate getEnd();
+
+    /**
+     * Returns the content to be sent to the telemetry service.
+     *
+     * This method is called periodically, once per content submission.
+     *
+     * @return
+     */
+    @Nonnull
+    public abstract JSONObject createContent();
+
+    public static ExtensionList<Telemetry> all() {
+        return ExtensionList.lookup(Telemetry.class);
+    }
+
+    @Extension
+    public static class TelemetryReporter extends AsyncPeriodicWork {
+
+        public TelemetryReporter() {
+            super("telemetry collection");
+        }
+
+        @Override
+        public long getRecurrencePeriod() {
+            return TimeUnit.HOURS.toMillis(24);
+        }
+
+        @Override
+        protected void execute(TaskListener listener) throws IOException, InterruptedException {
+            if (UsageStatistics.DISABLED || !Jenkins.getInstance().isUsageStatisticsCollected()) {
+                LOGGER.info("Collection of anonymous usage statistics is disabled, skipping telemetry collection and submission");
+                return;
+            }
+            Telemetry.all().forEach(telemetry -> {
+                if (telemetry.getStart().isAfter(LocalDate.now())) {
+                    LOGGER.config("Skipping telemetry for '" + telemetry.getId() + "' as it is configured to start later");
+                    return;
+                }
+                if (telemetry.getEnd().isBefore(LocalDate.now())) {
+                    LOGGER.config("Skipping telemetry for '" + telemetry.getId() + "' as it is configured to end in the past");
+                    return;
+                }
+
+                JSONObject data = new JSONObject();
+                try {
+                    data = telemetry.createContent();
+                } catch (Exception e) {
+                    LOGGER.log(Level.WARNING, "Failed to build telemetry content for: '" + telemetry.getId() + "'", e);
+                }
+
+                JSONObject wrappedData = new JSONObject();
+                wrappedData.put("type", telemetry.getId());
+                wrappedData.put("payload", data);
+                wrappedData.put("correlator", ExtensionList.lookupSingleton(Correlator.class).getCorrelationId());
+
+                try {
+                    URL url = new URL(ENDPOINT);
+                    URLConnection conn = ProxyConfiguration.open(url);
```

Comment:
should it be try-with-resources?

Reply:
Not closeable. Try-with-resources is a few lines further down.